File: BoardGamesSharp/Core/GameManager.cs
using BoardGamesSharp.Yahtzee;
using System;
using System.Collections.Generic;
using System.Diagnostics.Metrics;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace BoardGamesSharp.Core;

public class GameManager
{
    private List<Player> players = new List<Player>();

    public void Run()
    {
        Game currentGame = ChooseGame();

        if (currentGame == null) { Console.WriteLine("Не удалось запустить игру."); }
        else
        {
            Console.WriteLine("Запуск игры...\n");
            currentGame.Start();
        }


        Console.WriteLine("Текущие результаты после игры: ");
        foreach(Player player in players)
        {
            Console.WriteLine($"{player.Name}: {player.Wins}");
        }
    }

    public void AddPlayers()
    {
        int count = -1;
        while (count <= 0)
        {
            Console.Write("Введите количество игроков: ");
            string countStr = Console.ReadLine();
            if (!int.TryParse(countStr, out count) || count <= 0) { Console.WriteLine("Некорректный ввод."); }
        }

        players.Clear();
        for (int i = 0; i < count; i++)
        {
            Console.Write($"Имя игрока {i + 1}: ");
            string name = Console.ReadLine();
            players.Add(new Player(name));
        }
    }

    private Game ChooseGame()
    {
        Console.WriteLine("\nВыберите игру:");
        Console.WriteLine("1. Yahtzee");
        Console.WriteLine("2. (Пока не реализовано)");
        Console.WriteLine("0. Выход");

        int gameChoice = 0;
        Console.Write("Выберете игру: ");
        string gameChoiceStr = Console.ReadLine();
        if (!int.TryParse(gameChoiceStr, out gameChoice)) 
        { 
            Console.WriteLine("Некорректный ввод.");
        }

        switch (gameChoice)
        {
            case 0: return null;
            case 1: 
                return new YahtzeeGame(players);
            default: 
                Console.WriteLine("Пока не реализовано.");
                return ChooseGame();
        }
        return null;
    }

    public List<Player> GetPlayerList() { return players; }
}

File: BoardGames/GameManager.cpp
#include <iostream>
#include <vector>
#include "GameManager.hpp"
#include "YahtzeeGame.hpp"
#include "TestGame.hpp"
#include "StatsManager.hpp"


std::shared_ptr<Game> GameManager::Run() {
    auto currentGame = ChooseGame();

    if (!currentGame) {
        std::cout << "Выход из программы.\n";
        return nullptr;
    }
    else {
        totalGames++;
        currentGame->Run();
    }

    std::cout << "\nТекущие результаты после игры:\n";
    for (auto& p : players)
        std::cout << p->GetName() << ": " << p->GetStats() << "\n";

    return currentGame;
}

void GameManager::AddPlayers() {
    int count;
    std::cout << "Введите количество игроков: ";
    std::cin >> count;
    players.clear();

    for (int i = 0; i < count; i++) {
        std::string name;
        std::cout << "Имя игрока " << i + 1 << ": ";
        std::cin >> name;
        players.push_back(std::make_shared<Player>(name));
    }
}



std::unique_ptr<Game> GameManager::ChooseGame() {
    std::cout << "\nВыберите игру:\n";
    std::cout << "1. Yahtzee\n";
    std::cout << "2. Тест\n";
    std::cout << "3. (Пока не реализовано)\n";
    std::cout << "0. Выход\n";


    int gameChoice;
    std::cout << "\nВыберете игру: ";
    std::cin >> gameChoice;

    switch (gameChoice) {
    case 1:
        std::cout << "Хотите изменить число ходов? (Тек.: " << YahtzeeGame::GetHands() << ") ";
        char choice;
        std::cin >> choice;
        if (choice == 'y' || choice == 'Y') {
            int num = 0;
            while (num < 2 || num > 19) {
                std::cout << "Выберете число ходов (от 2 до 19): ";
                std::cin >> num;
            }
            YahtzeeGame::SetHands(num);
        }
        return std::make_unique<YahtzeeGame>(players, "Yahtzee");
    case 2:
        return std::make_unique<TestGame>(players, std::to_string(totalGames));
    case 0:
        return nullptr;
    default:
        std::cout << "Пока не реализовано.\n";
        return ChooseGame();
    }
}


